wrappers/hnyecho/echo.go

```diff
@@ -0,0 +1,81 @@
+package hnyecho
+
+import (
+	"github.com/jamietsao/beeline-go/wrappers/common"
+	"github.com/labstack/echo"
+)
+
+// EchoWrapper provides Honeycomb instrumentation for the Echo router via middleware
+type (
+	EchoWrapper struct {
+		handlerNames map[string]string
+	}
+)
+
+// New returns a new EchoWrapper struct
+func New() *EchoWrapper {
+	return &EchoWrapper{}
+}
+
+// Middleware returns an echo.MiddlewareFunc to be used with Echo.Use()
+func (e *EchoWrapper) Middleware() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			r := c.Request()
+			// get a new context with our trace from the request
+			ctx, span := common.StartSpanOrTraceFromHTTP(r)
+			defer span.Send()
+			// push the context with our trace and span on to the request
+			c.SetRequest(r.WithContext(ctx))
+
+			// get name of handler
+			handlerName := e.handlerName(c)
+			if handlerName == "" {
+				handlerName = "handler"
+			}
+			span.AddField("handler.name", handlerName)
+			span.AddField("name", handlerName)
+
+			// add field for each path param
+			for _, name := range c.ParamNames() {
+				span.AddField("app."+name, c.Param(name))
```

Comment:
One thing I wasn't sure about was the name of the field for both path params and query params (below).  In some of the example wrappers (e.g. gorilla and goji), analogous fields were named with `gorilla.vars.<var>` (or `goji.pat.<var>`).  I decided to go with `app.<var>` as I didn't want to couple the router implementation to the name.  I believe beeline uses `app.<var>` for any custom fields added so I figured this was a good default.

Thoughts?

Reply:
I agree the `gorilla.vars.<var>` and `goji.pat.<var>` is super annoying. I am leaning towards something like `route.vars.<var>` or some variant. I would like to keep the `app.*` namespace exclusively for fields added with `beeline.AddField()` so that it's clear "this stuff came from my manual instrumentation" vs. "this stuff came from a middleware in the beeline that's doing detection magic for me". 
I've been hesitant to change the gorilla and goji pattern because it's always annoying to change people's field names out from under them, and they're in the wild now.  And yet, it keeps coming up as a thing that's annoying people. I think the beeline may be headed for a major version bump in the near future; bringing sanity to middleware function names is a good thing to include in that.

Reply:
Makes sense.  Given that this Echo middleware will be new, I definitely want to use a forward looking naming convention.  I'm good with `route.vars.<var>` (or `path.vars.<var>`?).  Let me know what you prefer or think the next version will go with.

Reply:
@maplebed - I implemented all of your suggestions except this one.  Any preference for the prefix here?

Reply:
We're having hotly contested conversations about what's the best naming pattern. I think we're leaning towards `route`s rather than `path`s, and are on the fence about `param`s vs. `var`s. Sorry for the delay, but one more day would be lovely to get consensus.
(argument for params is that it echoes query params, which is (so far as I can find) the only one of these that actually has a canonical name. also echoes a function definition, which has parameters (though when called they're arguments?). Serious bikeshed time. One more day, promise.)

Reply:
Totally understand.  One of the two hardest things in computer science right?

Reply:
Ok. What we decided is that the `request.*` has things that are actually in the request, but that the routing engine, handlers, and variables are not actually part of the request. You can imagine a "has-a" relationship - a request _has a_ URL. A route will handle a request, but it's not the same thing, so routing gets its own top level namespace. A route does _have a_ handler and parameters. 

* `request.url` has the entire URL, unchanged as the user sent it (eg `/a/123?page=2`)
* `request.path` has the path portion of the URL (eg `/a/123`)
* `request.query` has the query portion of the URL (eg `page=2`)
* `route` has the matched portion of the path with variables eg `/a/:id/`
* `route.params.*` has all the variable names eg `route.params.id`
* `route.handler` has the name of the function defined by the route to handle the request

Does that sound reasonable?

Reply:
@maplebed - Yup, I like it.  I'll go ahead and make these changes.  Thanks!

Reply:
@maplebed  - I added the `route` fields as per your suggestion.  Should the `route` field for the matched portion of the path be `route.path` (similar to `request.path`)?